lms/services/grouping.py

```diff
@@ -80,49 +29,60 @@ def generate_authority_provided_id(
             tool_consumer_instance_guid, parent.lms_id, type_.value, lms_id
         )
 
-    def upsert_canvas_group(  # pylint: disable=too-many-arguments
+    def upsert_with_parent(  # pylint: disable=too-many-arguments
         self,
         tool_consumer_instance_guid,
-        context_id,
-        group_id,
-        group_name,
-        group_set_id,
+        lms_id,
+        lms_name,
+        parent: Grouping,
+        type_: Grouping.Type,
+        extra=None,
     ):
         """
-        Upserst a Grouping for a canvas group.
+        Upsert a Grouping generating the authority_provided_id based on its parent.
 
         :param tool_consumer_instance_guid: Tool consumer GUID
-        :param context_id: Course id the group is a part of
-        :param group_id: Canvas group id
-        :param group_name: The name of the group
-        :param group_set_id: Id of the canvas group set this group belongs to
+        :param lms_id: ID of this grouping on the LMS
+        :param lms_name: Name of the grouping on the LMS
+        :param parent: Parent of grouping
+        :param type_: Type of the grouping
+        :param extra: Any extra information to store linked to this grouping
         """
-
-        group_authority_provided_id = hashed_id(
-            tool_consumer_instance_guid, context_id, "canvas_group", group_id
-        )
-
-        course_authority_provided_id = hashed_id(
-            tool_consumer_instance_guid, context_id
+        return self._upsert(
+            Grouping(
+                application_instance=self._application_instance,
+                authority_provided_id=self.generate_authority_provided_id(
+                    tool_consumer_instance_guid, lms_id, parent, type_
+                ),
+                lms_id=lms_id,
+                lms_name=lms_name,
+                parent_id=parent.id,
+                extra=extra,
+                type=type_,
+            )
         )
 
-        course = self._course_service.get(course_authority_provided_id)
-
-        return self.upsert(
-            CanvasGroup(
-                application_instance=self._application_instance,
-                authority_provided_id=group_authority_provided_id,
-                lms_id=group_id,
-                lms_name=group_name,
-                parent_id=course.id,
-                extra={"group_set_id": group_set_id},
+    def _upsert(self, grouping):
+        db_grouping = (
+            self._db.query(Grouping)
+            .filter_by(
+                application_instance=grouping.application_instance,
+                authority_provided_id=grouping.authority_provided_id,
             )
+            .one_or_none()
         )
+        if not db_grouping:
+            self._db.add(grouping)
+        else:
+            # Update any fields that might have changed
+            db_grouping.lms_name = grouping.lms_name
+            db_grouping.extra = grouping.extra
+
+        return db_grouping or grouping
```

Comment:
I'm not sure that `_upsert()` really needs to be a separate method anymore since it's only ever called by the `upsert_with_parent()` method right above it, and if you inline it the code can be simpler. This passes all the tests, and it's particularly nice that there's only one `Grouping` object involved:

```python
class GroupingService:
    ...

    def upsert_with_parent(  # pylint: disable=too-many-arguments
        self,
        tool_consumer_instance_guid,
        lms_id,
        lms_name,
        parent: Grouping,
        type_: Grouping.Type,
        extra=None,
    ):
        authority_provided_id = self.generate_authority_provided_id(
            tool_consumer_instance_guid, lms_id, parent, type_
        )

        grouping = (
            self._db.query(Grouping)
            .filter_by(
                application_instance=self._application_instance,
                authority_provided_id=authority_provided_id,
            )
            .one_or_none()
        )

        if not grouping:
            grouping = Grouping(
                application_instance=self._application_instance,
                authority_provided_id=authority_provided_id,
                lms_id=lms_id,
                parent_id=parent.id,
                type=type_,
            )
            self._db.add(grouping)

        grouping.lms_name = lms_name
        grouping.extra = extra

        return grouping
```

Reply:
Inlined upsert, we might have to recover it later down the line but not for blackboard groups and we might need a slightly different method so this is better and cleaner for now.